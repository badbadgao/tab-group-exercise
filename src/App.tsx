import React from 'react';
import Tabs from 'components/Tabs';
import HomePage from 'pages/HomePage';
import AboutPage from 'pages/AboutPage';
import ProductsPage from 'pages/ProductsPage';
import PageNotFound from 'pages/PageNotFound';

import 'css/index.css';

import { Route, Switch, Redirect } from 'react-router-dom';
import { tabs, getTabId } from 'config/app.config';

const App = (): JSX.Element => {
  return (
    <div className="app">
      <Tabs tabs={tabs} />
      <Switch>
        <Route exact path="/home">
          <HomePage tabId={getTabId('/home')} />
        </Route>
        <Route exact path="/about">
          <AboutPage tabId={getTabId('/about')} />
        </Route>
        <Route exact path="/products">
          <ProductsPage tabId={getTabId('/products')} />
        </Route>
        <Route exact path="/">
          <Redirect to="/home" />
        </Route>
        <Route component={PageNotFound} />
      </Switch>
    </div>
  );
};

export default App;
