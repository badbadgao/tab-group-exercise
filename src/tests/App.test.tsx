import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import App from '../App';

test('check the whole app with 3 tabs', () => {
  const history = createMemoryHistory();
  const app = render(
    <Router history={history}>
      <App />
    </Router>,
  );

  // Snapshot test
  expect(app).toMatchSnapshot();

  // 1 element with the role 'tablist'
  const tabList = screen.getAllByRole('tablist');
  expect(tabList).toHaveLength(1);

  // 3 element with the role 'tablist'
  const tabs = screen.getAllByRole('tab');
  expect(tabs).toHaveLength(3);

  // Only the home tab is selected
  const activeTabs = screen.getAllByRole('tab', { selected: true });
  expect(activeTabs).toHaveLength(1);
  expect(activeTabs[0]).toHaveTextContent('Home');

  // Only products and about tab are not selected
  const inActiveTabs = screen.getAllByRole('tab', { selected: false });
  expect(inActiveTabs).toHaveLength(2);
  expect(inActiveTabs[0]).toHaveTextContent('Products');
  expect(inActiveTabs[1]).toHaveTextContent('About');

  // only the home tab panel is displaying
  const tabPanels = screen.getAllByRole('tabpanel');
  expect(tabPanels).toHaveLength(1);
  expect(tabPanels[0]).toHaveAttribute('aria-labelledby', 'tab-home');
  expect(tabPanels[0]).toHaveTextContent(/This is home page/i);
});

test('tab on click change', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <App />
    </Router>,
  );

  const activeTab = screen.getByRole('tab', { selected: true });
  expect(activeTab).toHaveTextContent('Home');
  const inActiveTabs = screen.getAllByRole('tab', { selected: false });
  expect(inActiveTabs).toHaveLength(2);
  const productsTab = inActiveTabs[0];
  const aboutTab = inActiveTabs[1];
  expect(productsTab).toHaveTextContent('Products');
  expect(aboutTab).toHaveTextContent('About');

  // click products tab
  fireEvent.click(productsTab);
  // expect the selected tab to be products tab
  const updatedActiveTabs1 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs1).toHaveLength(1);
  expect(updatedActiveTabs1[0]).toHaveTextContent('Products');
  // expect the products page displays
  const tabPanels1 = screen.getAllByRole('tabpanel');
  expect(tabPanels1).toHaveLength(1);
  expect(tabPanels1[0]).toHaveAttribute('aria-labelledby', 'tab-products');
  expect(tabPanels1[0]).toHaveTextContent(/This is products page/i);

  // click about tab
  fireEvent.click(aboutTab);
  // expect the selected tab to be about tab
  const updatedActiveTabs2 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs2).toHaveLength(1);
  expect(updatedActiveTabs2[0]).toHaveTextContent('About');
  // expect the about page displays
  const tabPanels2 = screen.getAllByRole('tabpanel');
  expect(tabPanels2).toHaveLength(1);
  expect(tabPanels2[0]).toHaveAttribute('aria-labelledby', 'tab-about');
  expect(tabPanels2[0]).toHaveTextContent(/This is about page/i);
});

test('tab keydown change - arrow right key', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <App />
    </Router>,
  );

  const tabList = screen.getAllByRole('tablist');
  expect(tabList).toHaveLength(1);

  const activeTab = screen.getByRole('tab', { selected: true });
  expect(activeTab).toHaveTextContent('Home');
  const inActiveTabs = screen.getAllByRole('tab', { selected: false });
  expect(inActiveTabs).toHaveLength(2);
  const productsTab = inActiveTabs[0];
  const aboutTab = inActiveTabs[1];
  expect(productsTab).toHaveTextContent('Products');
  expect(aboutTab).toHaveTextContent('About');

  // trigger arrow right key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowRight', code: 'ArrowRight' });
  // expect the selected tab to be products tab
  const updatedActiveTabs1 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs1).toHaveLength(1);
  expect(updatedActiveTabs1[0]).toHaveTextContent('Products');
  const tabPanels1 = screen.getAllByRole('tabpanel');
  expect(tabPanels1).toHaveLength(1);
  expect(tabPanels1[0]).toHaveAttribute('aria-labelledby', 'tab-products');
  expect(tabPanels1[0]).toHaveTextContent(/This is products page/i);

  // trigger arrow right key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowRight', code: 'ArrowRight' });
  // expect the selected tab to be about tab
  const updatedActiveTabs2 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs2).toHaveLength(1);
  expect(updatedActiveTabs2[0]).toHaveTextContent('About');
  const tabPanels2 = screen.getAllByRole('tabpanel');
  expect(tabPanels2).toHaveLength(1);
  expect(tabPanels2[0]).toHaveAttribute('aria-labelledby', 'tab-about');
  expect(tabPanels2[0]).toHaveTextContent(/This is about page/i);

  // trigger arrow right key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowRight', code: 'ArrowRight' });
  // expect the selected tab to be home tab
  const updatedActiveTabs3 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs3).toHaveLength(1);
  expect(updatedActiveTabs3[0]).toHaveTextContent('Home');
  const tabPanels3 = screen.getAllByRole('tabpanel');
  expect(tabPanels3).toHaveLength(1);
  expect(tabPanels3[0]).toHaveAttribute('aria-labelledby', 'tab-home');
  expect(tabPanels3[0]).toHaveTextContent(/This is home page/i);
});

test('tab keydown change - arrow left key', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <App />
    </Router>,
  );

  const tabList = screen.getAllByRole('tablist');
  expect(tabList).toHaveLength(1);

  const activeTab = screen.getByRole('tab', { selected: true });
  expect(activeTab).toHaveTextContent('Home');
  const inActiveTabs = screen.getAllByRole('tab', { selected: false });
  expect(inActiveTabs).toHaveLength(2);
  const productsTab = inActiveTabs[0];
  const aboutTab = inActiveTabs[1];
  expect(productsTab).toHaveTextContent('Products');
  expect(aboutTab).toHaveTextContent('About');

  // trigger arrow left key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowLeft', code: 'ArrowLeft' });
  // expect the selected tab to be about tab
  const tabPanels1 = screen.getAllByRole('tabpanel');
  expect(tabPanels1).toHaveLength(1);
  expect(tabPanels1[0]).toHaveAttribute('aria-labelledby', 'tab-about');
  expect(tabPanels1[0]).toHaveTextContent(/This is about page/i);
  const updatedActiveTabs1 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs1).toHaveLength(1);
  expect(updatedActiveTabs1[0]).toHaveTextContent('About');

  // trigger arrow left key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowLeft', code: 'ArrowLeft' });
  // expect the selected tab to be products tab
  const updatedActiveTabs2 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs2).toHaveLength(1);
  expect(updatedActiveTabs2[0]).toHaveTextContent('Products');
  const tabPanels2 = screen.getAllByRole('tabpanel');
  expect(tabPanels2).toHaveLength(1);
  expect(tabPanels2[0]).toHaveAttribute('aria-labelledby', 'tab-products');
  expect(tabPanels2[0]).toHaveTextContent(/This is products page/i);

  // trigger arrow left key down event
  fireEvent.keyDown(tabList[0], { key: 'ArrowLeft', code: 'ArrowLeft' });
  // expect the selected tab to be home tab
  const updatedActiveTabs3 = screen.getAllByRole('tab', { selected: true });
  expect(updatedActiveTabs3).toHaveLength(1);
  expect(updatedActiveTabs3[0]).toHaveTextContent('Home');
  const tabPanels3 = screen.getAllByRole('tabpanel');
  expect(tabPanels3).toHaveLength(1);
  expect(tabPanels3[0]).toHaveAttribute('aria-labelledby', 'tab-home');
  expect(tabPanels3[0]).toHaveTextContent(/This is home page/i);
});
