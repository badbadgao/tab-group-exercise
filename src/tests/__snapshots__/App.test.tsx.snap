// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the whole app with 3 tabs 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app"
      >
        <nav
          class="tabs"
        >
          <div
            aria-labelledby="nav-tablist"
            role="tablist"
          >
            <button
              aria-controls="tabpanel-home"
              aria-selected="true"
              id="tab-home"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span>
                Home
              </span>
            </button>
            <button
              aria-controls="tabpanel-products"
              aria-selected="false"
              id="tab-products"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <span>
                Products
              </span>
            </button>
            <button
              aria-controls="tabpanel-about"
              aria-selected="false"
              id="tab-about"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <span>
                About
              </span>
            </button>
          </div>
        </nav>
        <div
          aria-labelledby="tab-home"
          id="tabpanel-home"
          role="tabpanel"
          tabindex="0"
        >
          <h1>
            This is home page
          </h1>
          <p>
            Learn to use the accessibility semantics defined by the Accessible Rich Internet Application (ARIA) specification to create accessible web experiences. This guide describes how to apply accessibility semantics to common design patterns and widgets. It provides design patterns and functional examples complemented by in-depth guidance for fundamental practices.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app"
    >
      <nav
        class="tabs"
      >
        <div
          aria-labelledby="nav-tablist"
          role="tablist"
        >
          <button
            aria-controls="tabpanel-home"
            aria-selected="true"
            id="tab-home"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span>
              Home
            </span>
          </button>
          <button
            aria-controls="tabpanel-products"
            aria-selected="false"
            id="tab-products"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Products
            </span>
          </button>
          <button
            aria-controls="tabpanel-about"
            aria-selected="false"
            id="tab-about"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              About
            </span>
          </button>
        </div>
      </nav>
      <div
        aria-labelledby="tab-home"
        id="tabpanel-home"
        role="tabpanel"
        tabindex="0"
      >
        <h1>
          This is home page
        </h1>
        <p>
          Learn to use the accessibility semantics defined by the Accessible Rich Internet Application (ARIA) specification to create accessible web experiences. This guide describes how to apply accessibility semantics to common design patterns and widgets. It provides design patterns and functional examples complemented by in-depth guidance for fundamental practices.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
